controller/api/destination/watcher/profile_watcher.go

```diff
@@ -0,0 +1,234 @@
+	publisher, ok := pw.profiles[name]
+	pw.profilesMu.RUnlock()
+	if !ok {
+		return fmt.Errorf("cannot unsubscribe from unknown service [%s] ", name)
+	}
+	publisher.unsubscribe(listener)
+	return nil
+}
+
+func (pw *ProfileWatcher) addProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.Lock()
+	publisher, ok := pw.profiles[id]
+	if !ok {
+		publisher = pw.newProfilePublisher(id, profile)
+		pw.profiles[id] = publisher
+
+	}
+	pw.profilesMu.Unlock()
+
+	publisher.update(profile)
+}
+
+func (pw *ProfileWatcher) updateProfile(old interface{}, new interface{}) {
+	pw.addProfile(new)
+}
+
+func (pw *ProfileWatcher) deleteProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.RLock()
+	publisher, ok := pw.profiles[id]
+	pw.profilesMu.RUnlock()
+	if ok {
+		publisher.update(nil)
+	}
+}
+
+func (pw *ProfileWatcher) newProfilePublisher(id ProfileID, profile *sp.ServiceProfile) *profilePublisher {
+	return &profilePublisher{
+		profile:   profile,
+		listeners: make([]ProfileUpdateListener, 0),
+		log: pw.log.WithFields(logging.Fields{
+			"component": "profile-publisher",
+			"ns":        id.Namespace,
+			"profile":   id.Name,
+		}),
+	}
+}
+
+////////////////////////
+/// profilePublisher ///
+////////////////////////
+
+func (pp *profilePublisher) subscribe(listener ProfileUpdateListener) {
+	pp.mutex.Lock()
+	defer pp.mutex.Unlock()
+
+	pp.listeners = append(pp.listeners, listener)
+	listener.Update(pp.profile)
+}
+
+// unsubscribe returns true iff the listener was found and removed.
+// it also returns the number of listeners remaining after unsubscribing.
+func (pp *profilePublisher) unsubscribe(listener ProfileUpdateListener) {
+	pp.mutex.Lock()
+	defer pp.mutex.Unlock()
+
+	for i, item := range pp.listeners {
+		if item == listener {
+			// delete the item from the slice
+			n := len(pp.listeners)
+			pp.listeners[i] = pp.listeners[n-1]
+			pp.listeners[n-1] = nil
+			pp.listeners = pp.listeners[:n-1]
+			return
+		}
+	}
+}
+
+func (pp *profilePublisher) update(profile *sp.ServiceProfile) {
+	pp.mutex.Lock()
+	defer pp.mutex.Unlock()
+	pp.log.Debug("Updating profile")
+
+	pp.profile = profile
+	for _, listener := range pp.listeners {
+		listener.Update(profile)
+	}
+}
+
+////////////
+/// util ///
+////////////
+
+func nsFromToken(token string) string {
+	// ns:<namespace>
+	parts := strings.Split(token, ":")
+	if len(parts) == 2 && parts[0] == "ns" {
+		return parts[1]
+	}
+
+	return ""
+}
+
+func profileID(authority string, contextToken string) (ProfileID, error) {
+	host, _, err := getHostAndPort(authority)
+	if err != nil {
+		return ProfileID{}, err
+	}
+	service, _, err := GetServiceAndPort(authority)
+	if err != nil {
+		return ProfileID{}, err
+	}
+	name := ProfileID{
+		Name:      host,
+		Namespace: service.Name,
```

Comment:
Yes, and actually, I think this is a serious bug in the profile fallback behavior.  I'll add a test to fix the fact that there's no test coverage of this fallback logic.

We don't require context tokens because for each request we do two lookups: one with the context token (to look up the profile in the client's namespace) and one without the context token (to look up the profile in the server's namespace).